### ISSUE TITLE
Is it possible (and easy) to compile this source code?

### ISSUE THREAD

**Issue by byxor** [thug1src/thug#1]
I don't see any makefiles, and I'm assuming this code requires additional libraries that we don't have access to.

**Comment by ericwomer** [comment]
The real problem right now is the legality of even looking at this code. As this isn't released via the copyright owners.

**Comment by LWSS** [comment]
The way they have it set up is that certain folders will have subfolders for each platform.
Like in `Sys` they have NGC for Nintendo GameCube, NGPS for the playstation, XBox for the xbox, and Win32 for windows.

In each of these folders there are cpp and h files starting with `p_` for platform specific implementations.  

Unfortunately it is missing some of the Win32 folders, in big parts like Gfx, Music, SoundFX, Movies.

It looks possible to build one of the console versions with the right toolchain, but for PC, you would have to do some work.

**Comment by byxor** [comment]
Just write our own. Easy! :wink:

**Comment by ericwomer** [comment]
> Just write our own. Easy! wink

Are you talking a complete rewrite or using an existing engine?

**Comment by byxor** [comment]
I was only joking, but if the platform-dependent code isn't that big, it wouldn't be _too_ difficult to rewrite from scratch. I'm pretty sure CHC has done this before.

**Comment by ericwomer** [comment]
Its the legality of the issue, if we could get the current ip/source rights owners to sign off on it I would suggest taking out what we need, loading resources (models, images) and scripting (loading and running) and adding it to a compatible modern engine and go from there, if not, we would have to not look at the code and start from scratch to keep the lawers off our asses. I haven't looked much at the code, but from what I searched for online, its a renderware based game, like GTA and others, so I don't think its as complete as you think.

**Comment by LWSS** [comment]
@salamanderrake 
sounds like ur making excuses bub

` I haven't looked much at the code`

`I don't think its as complete as you think`

how can you make this conclusion :thinking:

**Comment by ericwomer** [comment]
Simple, because it was released by a studio, someone in the studio, or had access to the code, that didn't have rights to the code to the renderware engine its self, kind like how helium rain released their game code minus the UE4 engine its self and the 3d party stuff from the ue4 market. Also where is the documentation and the build scripts? They either had access to only one folder or they decided to only copy one folder.

**Comment by swift502** [comment]
To me, it feels very complete in terms of gameplay mechanics. We can now replicate the exact gameplay of Tony Hawk. We have the exact "magic values". And everything is very nicely commented. (In a Neversoft fashion)

https://github.com/thug1src/thug/blob/d8eb7147663d28c5cff3249a6df7d98e692741cb/Code/Gel/Music/ngc/bgm/bgm_play.c#L792-L797

The most interesting stuff seems to be in [SkaterCorePhysicsComponent.cpp](https://github.com/thug1src/thug/blob/master/Code/Sk/Components/SkaterCorePhysicsComponent.cpp), that's where we could never be sure until now. Few other SkaterComponents and I think you're done. Sipmly offload everything else (assets, rendering, etc.) onto a different engine.

I'll attempt at doing this but it is indeed a massive project to take on. I'll see how it goes for me. I must say though, I was very excited to see that about 90% of my predictions on how acid drops, spine transfers and verts work were correct. Feels good. :slightly_smiling_face: 

@thug1src @salamanderrake Btw when was this leaked? 2016? How did I miss it? And why thug of all games?

**Comment by ericwomer** [comment]
Commits on Feb 13, 2016  from what the commits say. Like I said before, my only concern is the legality of it all, its nice to have the code, but it would suck to start on this, just to have some piss-poor lawer jump up our asses and shut it down. I've seen it done to the wine project with the windows 2000 leaks and bits of that being pushed into trunk.

**Comment by swift502** [comment]
I was thinking about rewriting the important bits in a different language (C#, javascript, ...). I can't imagine someone connecting your polished rewrite and integration into something like Unity with this leaked code. I might be wrong though.

But looking further into the code, I'm not so sure any of this is a great idea.

Whenever you stop tracking a ramp, there's about 14 raycasts just to make sure you're really off a ramp.

https://github.com/thug1src/thug/blob/d8eb7147663d28c5cff3249a6df7d98e692741cb/Code/Sk/Components/SkaterCorePhysicsComponent.cpp#L5039-L5045

And when acid dropping / spine transfering, there's **80 raycasts** to find out if there's something you can transfer to.

https://github.com/thug1src/thug/blob/d8eb7147663d28c5cff3249a6df7d98e692741cb/Code/Sk/Components/SkaterCorePhysicsComponent.cpp#L3298-L3307

I thought we could easily port the key mechanics something like Unity physics, but the algorithms are incredibly primitive and rely on the optimized custom physics that they have going. And I think it would be very challenging to optimize it properly in a different engine.

In short, doing anything with this is probably a bad idea for several reasons. It's nice to explore and get inspired if you wanted to potentially make your own Tony Hawk game, and it would be nice to have the latest version (THPG) of these algorithms, but that's about it.

**Comment by ericwomer** [comment]
There are engines from Godot to UE4 that could be used.

**Comment by ericwomer** [comment]
@swift502 @LWSS @byxor It looks like who ever leaked this (if its even a leak or what) tried to updated it at least after 2011 (C++11) 	https://github.com/thug1src/thug/blob/d8eb7147663d28c5cff3249a6df7d98e692741cb/Code/Core/Debug/XBox/p_debug.cpp#L335 It looks like someone tried to updated the code, so this may not be vanilla or maybe, like in the code it suggests, may be for other versions of the games. There is also a main in the code, https://github.com/thug1src/thug/blob/d8eb7147663d28c5cff3249a6df7d98e692741cb/Code/Sk/Main.cpp#L245, so it suggests that the game is here, weather or not the renderware side of it or other deps are here is another question.

**Comment by LWSS** [comment]
@salamanderrake 
reinterpret_cast is older than C++11

I do know that they were using gcc, because I saw some `__attribute__` when I was trying to get it to compile.

**Comment by ericwomer** [comment]
> @salamanderrake
> reinterpret_cast is older than C++11
> 
> I do know that they were using gcc, because I saw some `__attribute__` when I was trying to get it to compile.

Yeah, I see that now, I thought all those *_cast were a c++11 thing.

**Comment by ericwomer** [comment]
Also we could just wait for the next TH inspired game from a different publisher, http://www.nintendolife.com/news/2018/02/tony_hawk_laments_the_fact_that_he_doesnt_own_the_pro_skater_brand

**Comment by byxor** [comment]
@swift502 @salamanderrake 

It's exciting to see this discussion suddenly blow up. I don't think pursuing compiling this will be worth it for anyone; It's a lot of work and most people would continue playing THUG Pro anyway.

Maybe someday someone will use this engine to build a full THPS6, but I think THUG Pro is the grand finale of THPS for the foreseeable future. However, don't let me deter you, this would be an exciting project that I could reap the sweet benefits from. :wink:

**Comment by ericwomer** [comment]
THUG Pro? THPS6? Or are you talking about the codewords for the other no PS games in the code its self?  It seems there are conflicting stories about THPS6, on one hand Tony Hawk was talking about continuing the franchise after activation in 20016, but yet again this year in feburary TH said he dosen't own the rights to the THPS brand, so I don't know exactly whats going on. No one would use this engine to make anymore THPS games as its renderer is based of the now defunct renderware library. Well thats not true, as they could pull the THPS code out and use a different rendering library, but thats what they did for the last game I believe.

**Comment by byxor** [comment]
@salamanderrake 
[THUG Pro](http://thugpro.com/) is a THUG2 mod. It released in ~2014, revived THPS and has had an expanding community ever since. It's the definitive way to play THPS. You can join [our discord server](https://discord.gg/nRXU8pH) if you're interested.

THPS6 doesn't exist.

**Comment by michael-fa** [comment]
doing anything with this game is probably (better to) just go and reverse the game you'd like to make better..
It's kinda weird for me actually - btw im not that experienced in doing that but I got a really long, hard time reversing it at first.

**Comment by byxor** [comment]
A giant chunk of the game is written in the QB scripts anyway (which aren't in this repo). There are QB decompilers and compilers out there (that don't work very well) that make that form of modding much more accessible than C++.

Even if we never recompile or rewrite the engine, we still have control over a large portion of content.

**Comment by michael-fa** [comment]
> A giant chunk of the game is written in the QB scripts anyway (which aren't in this repo). There are QB decompilers and compilers out there (that don't work very well) that make that form of modding much more accessible than C++.
> 
> Even if we never recompile or rewrite the engine, we still have control over a large portion of content.

yep. Thats what you really wanna do overall.. I only reversed all I need to access every qb related stuff - that made did it even tho the engine is weird for me to "understand"

**Comment by Cybertwip** [comment]
Has anyone successfully compiled to either platform?

Is the code complete? and if not, what is missing?

**Comment by michael-fa** [comment]
Game assets basically are all missing
And it would only allow one to see how the good stuff was made.

Balanced Guy ***@***.***> schrieb am Mo., 10. Mai 2021, 23:49:

**Comment by swift502** [comment]
@balancedwolf I believe it would be infinitely easier to just borrow the important bits of code for the movement, and all of the magic constants, but replace physics, rendering and UI with a modern day solution (Unity, Unreal), than to piece the code together and reimplement all the ancient or missing libraries and assets.

**Comment by Cybertwip** [comment]
I have the Xbox .xbx data files, they can be ripped directly from the original THUG Xbox game, I think this game in particular does not have data encryption, I will try to build an Xbox version, Just wanted to make sure this code is complete, and if not, there could be a restoration task long to be done

**Comment by Cybertwip** [comment]
Believing CMake would make it easier to generate solution files for this game in particular, Dev Kits will be needed also, UWP seems happy with anything you throw and it wouldn't be hard to translate between the old Xbox to Xbox One.

**Comment by Cybertwip** [comment]
I've forked the project, skatecareer.h is missing, must create, and possibly more files will be missing

You'll need DirectX 8
https://archive.org/download/dx8a_sdk/DX8a_SDK.exe 

For Win32 we will need to emulate the Xbox IDirect3DPalette8, which might be as well a texture

I'll keep going in a chance

https://github.com/balancedwolf/thug

Branch is Win32. All PS2 and GameCube source files were removed, so don't expect them

**Comment by Cybertwip** [comment]
I wonder if it's worth reconstructing all of this? Probably a few months or years to have it running properly.

**Comment by michael-fa** [comment]
There's way too much unmentioned problems or just missing code. 
Park editor or some other editors are missing.
I contacted Mick West, and he told me some stuff thats missing, and what "kind of" time/state this is from.

Doing anything, I guess best is to recycle bits of the code.
Better; For anyone expierienced... I follow Mick for game dev stuff, he posted tons of stuff online just go search him.
He explained how to handle game input and process all other stuff in the best way that there's a natural feeling.
Something along the lines about MainLoop - Controller Input -> Game Logic -> Graphics Loop or whatever.

**Comment by LWSS** [comment]
> There's way too much unmentioned problems or just missing code.
> Park editor or some other editors are missing.
> I contacted Mick West, and he told me some stuff thats missing, and what "kind of" time/state this is from.
> 
> Doing anything, I guess best is to recycle bits of the code.
> Better; For anyone expierienced... I follow Mick for game dev stuff, he posted tons of stuff online just go search him.
> He explained how to handle game input and process all other stuff in the best way that there's a natural feeling.
> Something along the lines about MainLoop - Controller Input -> Game Logic -> Graphics Loop or whatever.

Can you post it in here? I just see UFO schizo postings on his twitter

**Comment by huckleberrypie** [comment]
> > There's way too much unmentioned problems or just missing code.
> > Park editor or some other editors are missing.
> > I contacted Mick West, and he told me some stuff thats missing, and what "kind of" time/state this is from.
> > Doing anything, I guess best is to recycle bits of the code.
> > Better; For anyone expierienced... I follow Mick for game dev stuff, he posted tons of stuff online just go search him.
> > He explained how to handle game input and process all other stuff in the best way that there's a natural feeling.
> > Something along the lines about MainLoop - Controller Input -> Game Logic -> Graphics Loop or whatever.
> 
> Can you post it in here? I just see UFO schizo postings on his twitter

TBF the schizo postings you speak of is actually debunkings of his as he took up a second career as a skeptic.

**Comment by LWSS** [comment]
> > > There's way too much unmentioned problems or just missing code.
> > > Park editor or some other editors are missing.
> > > I contacted Mick West, and he told me some stuff thats missing, and what "kind of" time/state this is from.
> > > Doing anything, I guess best is to recycle bits of the code.
> > > Better; For anyone expierienced... I follow Mick for game dev stuff, he posted tons of stuff online just go search him.
> > > He explained how to handle game input and process all other stuff in the best way that there's a natural feeling.
> > > Something along the lines about MainLoop - Controller Input -> Game Logic -> Graphics Loop or whatever.
> > 
> > 
> > Can you post it in here? I just see UFO schizo postings on his twitter
> 
> TBF the schizo postings you speak of is actually debunkings of his as he took up a second career as a skeptic.

Being a 'skeptic' just sounds like a convenient excuse to be obsessed with cryptids/conspiracies/etc and avoid the usual negative stigma.

**Comment by LWSS** [comment]
btw multiple people have compiled this source code by now, this issue should be closed
